### PR TITLE
Feature 1926 gridstat openmp

### LINKED ISSUE #1926: Enhance Grid-Stat to use OpenMP for efficient computation of neighborhood statistics by setting $OMP_NUM_THREADS
## Describe the New Feature ##

OpenMP provides a simple way of adding shared-memory (on-node) parallelism to serial code.  It can be added incrementally, and can provide quick wins when a code has a few dominant hotspots.

The `fractional_coverage` routine in `data_plane_util.cc` contains a loop-nest representing one such hotspot. We aim to parallelize this loop with OpenMP. This would distribute loop iterations across multiple cores and pull down the overall time spent in GridStat.

OMP is directive-based. The directives begin with the comment character (`!`) and hence are ignored by the compiler if OpenMP support is not requested. To instruct the compiler to interpret OMP directives, an additional flag is normally required. This varies between compilers. With `gfortran` the flag is `-fopenmp`.

The number of threads can be controlled programmatically. However, it is sufficient to set environment variable `OMP_NUM_THREADS` at runtime.

### Acceptance Testing ###
*List input data types and sources.*

The code should give identical answers with 1 vs. 2+ threads. A 1-thread vs. 2-thread result comparison would be needed. Small differences can sometimes appear with/without OpenMP compiled in (even with 1 thread only) because serially optimising compilers sometimes make different choices when the OpenMP flag is present. I don't think that's a reason to omit an OpenMP vs. no-OpenMP test, though; just something to be aware of. It is certainly  possible to introduce bugs with OpenMP and only one thread, so OMP/no-OMP tests would be wise.

### Time Estimate ###
*Estimate the amount of work required here.*
*Issues should represent approximately 1 to 3 days of work.*

### Sub-Issues ###
Consider breaking the new feature down into sub-issues.
- [ ] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
2799991

### Assignee ###
- [ ] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [ ] Select **component(s)**
- [ ] Select **priority**
- [ ] Select **requestor(s)**

### Projects and Milestone ###
- [ ] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED PROJECT ASSIGNMENT** label
- [ ] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [ ] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)

## New Feature Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Linked issues**
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by mo-mglover**
## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>
      Scaling behaviour plots: given in slides linked of issue #1926.  Differencing of output stat files did not show any differences, and none are expected.  If differences - particularly transient differences - are observed, this would be symptomatic of a race condition (a bug).

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>
       OpenMP activated by setting `MET_OPENMP=true`, or equivalently to some random string. (Similar to `MET_DEVELOPMENT` in that regard.)  At runtime, the user sets environment variable `OMP_NUM_THREADS=<# threads>`.   If `OMP_NUM_THREADS` is left unset, then the behaviour is platform-dependent; it will likely default to the number of logical cores on the system (not 1).  

- [ ] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes or No]**

- [ ] Do these changes include sufficient testing updates? **[Yes or No]**

- [ ] Will this PR result in changes to the test suite? **[Yes or No]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

- [ ] Please complete this pull request review by **[Fill in date]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [ ] After submitting the PR, select **Linked issues** with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

**Comment by mo-mglover** [comment]
Opened the pull request anticipating discussion and a number of iterations.

I'm aware of a couple of areas that will need tweaking:

  * Documentation
  * Automatic tests

I've found the test suite config. files, but don't know how to modify them appropriately.  Essentially, we would want to run the same tests on 1 and 2 threads, and check that the results are identical. 

There's a further potential issue around the `configure.ac` file and how best to add in the relevant OpenMP compiler flag.   While there does seem to be some movement towards standardising around `-fopenmp`, this is not true in general today. GNU: `-fopenmp`; Intel: `-qopenmp`;  Cray: `-homp`.

**Comment by JohnHalleyGotway** [comment]
Thanks for submitting this PR Maff. FYI, I did the following:
- assigned to myself for review
- set Projects = MET-10.1.0-beta5
- set Milestone = MET 10.1.0
- set Linked issues = dtcenter/MET#1926

I'm starting by running a regression test to compare feature_1926_gridstat_openmp to the develop branch. Assuming that flags no diffs, then I'll try compiling WITH openmp enabled and confirm that the output remains unchanged.

I agree with the additional work you listed in this [comment](https://github.com/dtcenter/MET/pull/1977#issuecomment-974126890).

File: met/configure.ac [hunk 1]
```diff
@@ -1067,9 +1068,17 @@ AC_SUBST(FC_LIBS, [-lgfortran])
 # For Intel compilers, "-g" slows down runtimes considerably (MET #1778).
 # For development, retain the "-g" option. Otherwise, discard it.
 
-AM_COND_IF([ENABLE_DEVELOPMENT], [: ${CFLAGS="-g -O2"}], [: ${CFLAGS="-O2"}])
-AM_COND_IF([ENABLE_DEVELOPMENT], [: ${CXXFLAGS="-g"}  ], [: ${CXXFLAGS=""}] )
-AM_COND_IF([ENABLE_DEVELOPMENT], [: ${FFLAGS="-g -O2"}], [: ${FFLAGS="-O2"}])
+AM_COND_IF([ENABLE_DEVELOPMENT], [: ${CFLAGS_OPT="-g -O2"}], [: ${CFLAGS_OPT="-O2"}])
+AM_COND_IF([ENABLE_DEVELOPMENT], [: ${CXXFLAGS_OPT="-g"}  ], [: ${CXXFLAGS_OPT="-O3"}] )
```
Comment (JohnHalleyGotway) [inline met/configure.ac]: Was this change intentional... changing ${CXXFLAGS=""} to ${CXXFLAGS_OPT="-03"}?
I believe that when in "development" mode, we want to disable optimization for easier debugging.

Comment (mo-mglover) [reply met/configure.ac]: Good spot! No, it wasn't intentional. Will put that right and push up the correction.

File: met/src/tools/core/grid_stat/grid_stat.cc [hunk 2]
```diff
@@ -185,6 +187,15 @@ static bool read_data_plane(VarInfo* info, DataPlane& dp, Met2dDataFile* mtddf,
 
 int main(int argc, char *argv[]) {
 
+#ifdef _OPENMP
+#pragma omp parallel
+#pragma omp single
+  {
+    // Report number of threads if compiled with OpenMP
+    std::cout << "Running on " << omp_get_num_threads() << " threads.\n";
```
Comment (JohnHalleyGotway) [inline met/src/tools/core/grid_stat/grid_stat.cc]:
```suggestion
    mlog << Debug(2) << "Running on " << omp_get_num_threads() << " threads.\n";
```
Recommend routing log messages through the global logging class instance (mlog) instead of cout.

**Comment by JohnHalleyGotway** [comment]
@jprestop I added you as a reviewer of this PR as well since it impacts how the MET is compiled.
Note from Maff's [comment above](https://github.com/dtcenter/MET/pull/1977#issuecomment-974126890) the OpenMP compiler flag to used is compiler-dependent:
   GNU: -fopenmp; Intel: -qopenmp; Cray: -homp
Maff's current changes in configure.ac use -fopenmp.

You'll likely want to update the MET compilation script to include:
ENABLE_OPENMP and CFLAGS_OMP/CXXFLAGS_OMP/FFLAGS_OMP

**Comment by JohnHalleyGotway** [comment]
The unit_gen_vx_mask.xml tests failed for feature_1926_gridstat_openmp:
```
ERROR  : 
ERROR  : Out of memory reading GRIB2 data!  Exiting!
ERROR  : Check that MET and the GRIB2C library were compiled consistently, either with or without the -D__64BIT__ flag.
ERROR  : 
ERROR: /d1/projects/MET/MET_pull_requests/met-10.1.0/met-10.1.0_beta5/feature_1926/MET-feature_1926_gridstat_openmp/test/perl/unit.pl unit_gen_vx_mask.xml failed.
ERROR: Command returned with non-zero status (1): test/bin/unit_test.sh
```
The failure is caused by differences in the compiler flags:
```
egrep "CFLAGS=|CXXFLAGS=|FFLAGS=" MET-develop/met/config.log
CFLAGS='-g -DUNDERSCORE -fPIC -D__64BIT__'
CXXFLAGS='-g -DUNDERSCORE -fPIC -D__64BIT__'
FFLAGS='-g -O2'
```
versus the feature branch:
```
egrep "CFLAGS=|CXXFLAGS=|FFLAGS=" MET-feature_1926_gridstat_openmp/met/config.log
CFLAGS='-g -O2 '
CXXFLAGS='-g '
FFLAGS='-g -O2 '
```
The CFLAGS env var was set when configure was called:
```
ac_cv_env_CFLAGS_set=set
ac_cv_env_CFLAGS_value='-g -DUNDERSCORE -fPIC -D__64BIT__'
```
However configure.ac is overriding the input values starting at this line:
https://github.com/dtcenter/MET/blob/3ff12fe976a738c9f96f20ddc424a87eaf3e3f0b/met/configure.ac#L1079

Trouble is I don't fully understand the automake syntax used on this line:
https://github.com/dtcenter/MET/blob/4e28ae7f11b90e0cac989e53ee7357ab63d2c6c9/met/configure.ac#L1070

Perhaps that only sets CFLAGS is they are not already set?

I recommend we handle the openmp options similar to how the GRIB2 and Python options are handled. Consistent with other optional compilation options, I recommend we use an environment variable named "MET_OPENMP", which if set, triggers ENABLE_OPENMP to be set in configure.ac.

I'll make some changes to configure.ac directly in the feature_1926_gridstat_openmp branch. If it breaks everything, we can always back them out.

**Comment by mo-mglover** [comment]
Modified system for setting `-fopenmp` is much better than the way I'd done it. I've cherry-picked that and merged into this branch. Also fixed the linking problem (`-fopenmp` is also needed by the linker); that takes away the need to add `-lgomp` by hand for the GNU compiler.

File: met/src/tools/core/grid_stat/grid_stat.cc (hunk 2 above)
Comment (mo-mglover) [reply met/src/tools/core/grid_stat/grid_stat.cc]: Happy to do.  I'd first like to check that's indeed what we want to do, though.  My thinking is: it's valuable to have information about the runtime configuration in the main "stdout" file, so that we can cross-reference it with, say, execution time or even just determine whether the executable contains OpenMP.  I think it needs to be more than a debug message.  With those things in mind, what would be the best way to code it?

Comment (JohnHalleyGotway) [reply met/src/tools/core/grid_stat/grid_stat.cc]: @mo-mglover very, very few of the log messages written by MET go to stdout. One exception is a rather annoying message from GNU Scientific Library when we create a random number generator and can't get it to write anywhere other than stdout. Any error messages occurring before the global logger is instantiated must also be written to stdout. But in general, if we can log via the logger, we do. The logging class does in fact write to stdout, as well as redirect log messages to a log file, if so requested using the "-log filename" command line option. Each call to the logger starts with either Error, Warning, or Debug(n) where n is the logging level. If it's useful, we could add other categories of log messages for performance. But we'd need to decide the rules for if/when those log messages should be enabled/disabled. Right now the user can set "-v log_level" on the command line (with a default of 2). If the log message level in the code <= log_level, it gets printed to stdout and the log file (if one is specified).

I should mention that Warning and Error log messages are always printed, regardless of the user-defined log level.

**Comment by jprestop** [comment]
@JohnHalleyGotway 
Is there anything I need to test before this PR is approved? I see that you stated in an earlier comment:

> I'm starting by running a regression test to compare feature_1926_gridstat_openmp to the develop branch. Assuming that flags no diffs, then I'll try compiling WITH openmp enabled and confirm that the output remains unchanged.

So I wasn't sure if you needed me to do any further testing or not.

I also wanted to make sure that I understand what I need to do with the compile script.  It looks like an --enable-openmp will need to be added in the same manner as we either compile or don’t compile with --enable-grib2, but do I also need to add the following compilation flags for: 
GNU: -fopenmp; 
Intel: -qopenmp;
Cray: -homp
or is that handled in the autoconf setup if --enable-openmp is selected?

**Comment by JohnHalleyGotway** [comment]
@mo-mglover when $OMP_NUM_THREADS is not set, grid_stat runs on my Mac using 12 threads. I presume it's using all *available* threads by default. But I don't know how that is determined. I was expecting it to run on a single thread by default. The latter default seems a little "safer" so that leaving $OMP_NUM_THREADS unset yields the exact same behavior as previous versions.

But is there some standard here? How do you think this should be handled?

**Comment by JohnHalleyGotway** [comment]
@mo-mglover wanted to give you an update. @jprestop and I like the idea of using the existing AC_OPENMP() macro provided by autoconf. For GNU at least, it's smart enough to define the "-fopenmp" compiler flag without us needing to explicitly tell it. This compiles openmp support by default which is fine.

I pushed up that change to configure.ac, was able to remove some silly things that I'd added to the scripts directory, and updated installation.rst, replacing the --enable-openmp discussion with one about --disable-openmp.

@jprestop will test to make sure that the AC_OPENMP() macro works well with the Intel compilers. Fingers crossed.

The question remains of what to do when OMP_NUM_THREADS is not set. I think using a single thread would be most intuitive, but am open to suggestions.

**Comment by mo-mglover** [comment]
Pushed up a further change to the OpenMP directives in 9594505.    Using `default(none)`  to force the need for all variables to appear in data sharing clauses explicitly, such as `private` or `shared`.  Exception (C++): those declared inside the parallel region, which are automatically private.

**Comment by mo-mglover** [comment]
@JohnHalleyGotway  I like the  `AC_OPENMP` approach.  

You're right that the number of threads will default to the maximum number of hardware threads available on the system, not 1.  Running on 1 thread is not quite the same as no threading at all, due to `private` variables being uninitialized by default on entering a parallel region (no guarantees their value is valid).  In bug-free code, this won't be an issue.   Also, compiler optimizations may be subtly different between OpenMP-compiled code and non-OMP-compiled.

**Comment by mo-mglover** [comment]
@JohnHalleyGotway     In terms of defaulting the number of threads to 1 when `OMP_NUM_THREADS` is unset, we can either do this via external scripting or handle it in the program itself.  In order to not create a requirement for external scripting, and to enable writing a helpful message when OpenMP is disabled, I've added some new code to do this.

Noting that the `fractional_coverage` routine is called by other top-level tools,  I've added calls to `init_openmp()` in those, too.  

I'm wondering whether some of the new log messages should be warnings rather than debug messages.  Will sleep on that, or perhaps you have a view?

File: met/src/basic/vx_util/handle_openmp.cc [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+void init_openmp() {
+
+#ifdef _OPENMP
+
+  // If OMP_NUM_THREADS was not set, use the OpenMP API to set the thread count
+  // to 1 thread only.
+  const char* env_omp_num_threads = std::getenv("OMP_NUM_THREADS");
+  if (!env_omp_num_threads) {
+    mlog << Debug(2) << "OMP_NUM_THREADS is not set." 
+     << " Defaulting to 1 thread. \n";
```
Comment (JohnHalleyGotway) [inline met/src/basic/vx_util/handle_openmp.cc]:
```suggestion
         << " Defaulting to 1 thread. \n";
```
Just lining code up.

File: met/src/basic/vx_util/handle_openmp.cc [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    omp_set_num_threads(1);
+  }
+
+#pragma omp parallel
+#pragma omp single
+  {
+     mlog << Debug(2) << "OpenMP running on " 
+       << omp_get_num_threads() << " thread(s).\n";
```
Comment (JohnHalleyGotway) [inline met/src/basic/vx_util/handle_openmp.cc:40]:
```suggestion
          << omp_get_num_threads() << " thread(s).\n";
```
Just lining code up.

File: met/src/basic/vx_util/Makefile.am [hunk 5]
```diff
@@ -69,6 +69,7 @@ libvx_util_a_SOURCES = ascii_table.cc ascii_table.h \
                GridOffset.h GridOffset.cc \
                observation.h observation.cc \
                stat_column_defs.h \
+	       handle_openmp.h handle_openmp.cc \
```
Comment (JohnHalleyGotway) [inline met/src/basic/vx_util/Makefile.am]:
```suggestion
               handle_openmp.h handle_openmp.cc \
```
Just lining things up.

**Comment by JohnHalleyGotway** [comment]
@mo-mglover, adding init_openmp() to the vx_util library is an excellent idea! Thanks for doing that. That'll make it much easier to add openmp support to other applications.

As for log message levels, I do think that DEBUG(2) is appropriate. For users running via the METplus wrappers, we can make sure OMP_NUM_THREADS is set there.
 
I had considered whether we'd want to provide a command line option to override the OMP_NUM_THREADS environment variable. But when running MET on the same platform, you'd probably want OMP_NUM_THREADS set to the same value for each call. If a need arises if the future, we could always add that later.

@jprestop is testing the compilation of the feature branch with Intel compilers to confirm that the autoconf settings pick the right compilation flags.

**Comment by jprestop** [comment]
@JohnHalleyGotway 

I was able to get a successful build of the MET feature_1926_gridstat_openmp branch on kiowa at /d1/projects/MET/MET_pull_requests/met-10.1.0/met-10.1.0_beta5/feature_1926_intel, however, "-qopenmp" was NOT used but rather "-fopenmp" was used:
> checking for /usr/local/intel-18.0.3/bin/icc option to support OpenMP... -fopenmp

I sent the output to build.log in that directory and can see that -fopenmp was used throughout.  It looks like we need to revisit this issue.

**Comment by jprestop** [comment]
@mo-mglover 

@JohnHalleyGotway  and I reviewed the compilation of openmpi with the Intel compilers and reviewed the flag selection for PGI.

For Intel, it seems that -fopenmp is the same as using -qopenmp.  Running "icc --help" shows:
```
-qopenmp  enable the compiler to generate multi-threaded code based on the
          OpenMP* directives (same as -fopenmp)
```
from the config.log file for compiling MET we see that -fopenmp was chosen:
```
checking for /usr/local/intel-18.0.3/bin/icc option to support OpenMP... -fopenmp 
```
For PGI, it looks as if -mp was the chosen flag, which is expected. In config.log:
```
OPENMP_CFLAGS='-mp'    
```
Based on these tests, we confirm that the autoconf settings pick the right compilation flags for the GNU, Intel, and PGI compilers.

@JohnHalleyGotway and @jprestop  Need to add documentation on setting the OMP_NUM_THREADS runtime environment variable.

**Comment by JohnHalleyGotway** [comment]
Feedback in the [NOAA User's Telecon](https://docs.google.com/document/d/12crSdlVBu4RV_K4_27fOgCDQKxsO7kV_0DqBVZtOePw/edit) on 12/13/2021 recommended that we:
1. Leave the Debug(2) message level in place but make it a bit longer:
```
DEBUG 2: OMP_NUM_THREADS is not set. Defaulting to 1 thread. Recommend setting OMP_NUM_THREADS for faster runtimes.
```
2. Add a new section to the MET User's Guide about the use of OpenMP, describing OMP_NUM_THREADS and listing the portions of the code that are parallelized.

Add a new 4.1 section for "Runtime Environment Variables" and create 4.1.1, 4.1.2, and 4.1.3 to discuss MET_OBS_ERROR_TABLE, MET_GRIB_TABLES, and OMP_NUM_THREADS.

@jprestop will make these changes and @JohnHalleyGotway will review/update.

**Comment by mo-mglover** [comment]
@JohnHalleyGotway  I've tackled (1) and (2), and pushed up the changes, but I've haven't been able to "compile" the RST. (I'm getting errors that seem to be related to the sphinx setup at my end, rather than the RST itself.) 

I wasn't sure quite what was involved with the last paragraph. I've just noticed that @jprestop is tagged on that comment.

**Comment by JohnHalleyGotway** [comment]
Excellent @mo-mglover! Thanks for making these changes. @jprestop is the lead on documentation across all the METplus components. In the MET development meeting today, we discussed how/where to incorporate the OpenMP details into the User's Guide. So the next step is @jprestop integrating the details you provided in openmp.rst. After that, we'll be ready to merge this PR. Thanks for your work on this and patience.

**Comment by jprestop** [comment]
Thank you @mo-mglover for all of your work on this and for writing up the documentation as well.  It is very much appreciated.  @JohnHalleyGotway, I added the documentation to the config_options.rst file in a new section called "Runtime Environment Variables" as we discussed.  There are two things to note:
1.  In addition to MET_OBS_ERROR_TABLE, MET_GRIB_TABLES, and OMP_NUM_THREADS, there was also MET_BASE, so that is listed in 4.1.1.  Please let me know if that doesn't belong there.
2. Unfortunately, the Sphinx documentation for MET does not currently follow the standard (see below). As a result, the sections under OMP_NUM_THREADS "Introduction", "Setting the number of threads", "Which code is parallelized?", and "Thread Binding" are simply bolded.  This work is for another time, however.  I created a [new issue](https://github.com/dtcenter/MET/issues/1998) for this task.

> https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html
> \# with overline, for parts
> \* with overline, for chapters
> =, for sections
> -, for subsections
> ^, for subsubsections
> “, for paragraphs

I believe everything is in order for your final review, @JohnHalleyGotway.

**Review by jprestop — COMMENTED** [review]
I have added the documentation to the config_options.rst file in a new section called "Runtime Environment Variables" as we discussed.  In addition to MET_OBS_ERROR_TABLE, MET_GRIB_TABLES, and OMP_NUM_THREADS, there was also MET_BASE, so that is listed in 4.1.1. Please let me know if that doesn't belong there.  @JohnHalleyGotway I believe everything is in order for your final review.

**Review by JohnHalleyGotway — APPROVED** [review]
I approve of these changes.

On 1/12/22, I merged the latest changes from the develop branch into this feature branch and reran the regression test. Wanted to make sure there were no conflicts with recent development. The test flagged several differences but all of them are related to other development for which we have not yet updated the reference version.

Details can be found in kiowa:/d1/projects/MET/MET_pull_requests/met-10.1.0/met-10.1.0_beta5/feature_1926

```
egrep -i "file1:|file2:|ERROR:" test_regression.log | egrep -B 2 ERROR:
```

1. These files are added by this PR, generated using 2 processors.
```
ERROR: folder MET-develop/test_output missing 13 files
    grid_stat/grid_stat_GRIB2_NAM_RTMA_NP2_120000L_20120409_120000V_cnt.txt 
    grid_stat/grid_stat_GRIB2_NAM_RTMA_NP2_120000L_20120409_120000V_ctc.txt 
    grid_stat/grid_stat_GRIB2_NAM_RTMA_NP2_120000L_20120409_120000V_cts.txt 
    grid_stat/grid_stat_GRIB2_NAM_RTMA_NP2_120000L_20120409_120000V_eclv.txt 
    grid_stat/grid_stat_GRIB2_NAM_RTMA_NP2_120000L_20120409_120000V_fho.txt 
    grid_stat/grid_stat_GRIB2_NAM_RTMA_NP2_120000L_20120409_120000V_nbrcnt.txt 
    grid_stat/grid_stat_GRIB2_NAM_RTMA_NP2_120000L_20120409_120000V_nbrctc.txt 
    grid_stat/grid_stat_GRIB2_NAM_RTMA_NP2_120000L_20120409_120000V_nbrcts.txt 
    grid_stat/grid_stat_GRIB2_NAM_RTMA_NP2_120000L_20120409_120000V_pairs.nc 
    grid_stat/grid_stat_GRIB2_NAM_RTMA_NP2_120000L_20120409_120000V_sl1l2.txt 
    grid_stat/grid_stat_GRIB2_NAM_RTMA_NP2_120000L_20120409_120000V.stat 
    grid_stat/grid_stat_GRIB2_NAM_RTMA_NP2_120000L_20120409_120000V_vcnt.txt 
    grid_stat/grid_stat_GRIB2_NAM_RTMA_NP2_120000L_20120409_120000V_vl1l2.txt 
```

2. All of these many, many changes are caused by a small change in this PR that @JohnHalleyGotway made, switching "min 1" to "min 3". That causes INTERP_MTHD = MIN_SQAURE with INTERP_PNTS = 9 to appear in the output, as expected.
```
file1: MET-feature_1926_gridstat_openmp/test_output/grid_stat/grid_stat_GRIB2_NAM_RTMA_120000L_20120409_120000V_cnt.txt
file2: MET-develop/test_output/grid_stat/grid_stat_GRIB2_NAM_RTMA_120000L_20120409_120000V_cnt.txt
ERROR: header information mismatch in column INTERP_MTHD
ERROR: failed tests for CNT: hdr
file1: MET-feature_1926_gridstat_openmp/test_output/grid_stat/grid_stat_GRIB2_NAM_RTMA_120000L_20120409_120000V_ctc.txt
file2: MET-develop/test_output/grid_stat/grid_stat_GRIB2_NAM_RTMA_120000L_20120409_120000V_ctc.txt
ERROR: header information mismatch in column INTERP_MTHD
ERROR: failed tests for CTC: hdr
file1: MET-feature_1926_gridstat_openmp/test_output/grid_stat/grid_stat_GRIB2_NAM_RTMA_120000L_20120409_120000V_cts.txt
file2: MET-develop/test_output/grid_stat/grid_stat_GRIB2_NAM_RTMA_120000L_20120409_120000V_cts.txt
ERROR: header information mismatch in column INTERP_MTHD
ERROR: failed tests for CTS: hdr
file1: MET-feature_1926_gridstat_openmp/test_output/grid_stat/grid_stat_GRIB2_NAM_RTMA_120000L_20120409_120000V_eclv.txt
file2: MET-develop/test_output/grid_stat/grid_stat_GRIB2_NAM_RTMA_120000L_20120409_120000V_eclv.txt
ERROR: header information mismatch in column INTERP_MTHD
ERROR: failed tests for ECLV#66: hdr
file1: MET-feature_1926_gridstat_openmp/test_output/grid_stat/grid_stat_GRIB2_NAM_RTMA_120000L_20120409_120000V_fho.txt
file2: MET-develop/test_output/grid_stat/grid_stat_GRIB2_NAM_RTMA_120000L_20120409_120000V_fho.txt
ERROR: header information mismatch in column INTERP_MTHD
ERROR: failed tests for FHO: hdr
--
file1: MET-feature_1926_gridstat_openmp/test_output/grid_stat/grid_stat_GRIB2_NAM_RTMA_120000L_20120409_120000V_pairs.nc
file2: MET-develop/test_output/grid_stat/grid_stat_GRIB2_NAM_RTMA_120000L_20120409_120000V_pairs.nc
ERROR: NetCDF headers differ:
file1: MET-feature_1926_gridstat_openmp/test_output/grid_stat/grid_stat_GRIB2_NAM_RTMA_120000L_20120409_120000V_sl1l2.txt
file2: MET-develop/test_output/grid_stat/grid_stat_GRIB2_NAM_RTMA_120000L_20120409_120000V_sl1l2.txt
ERROR: header information mismatch in column INTERP_MTHD
ERROR: failed tests for SL1L2: hdr
file1: MET-feature_1926_gridstat_openmp/test_output/grid_stat/grid_stat_GRIB2_NAM_RTMA_120000L_20120409_120000V.stat
file2: MET-develop/test_output/grid_stat/grid_stat_GRIB2_NAM_RTMA_120000L_20120409_120000V.stat
ERROR: header information mismatch in column INTERP_MTHD
ERROR: header information mismatch in column INTERP_MTHD
ERROR: header information mismatch in column INTERP_MTHD
ERROR: header information mismatch in column INTERP_MTHD
ERROR: header information mismatch in column INTERP_MTHD
ERROR: header information mismatch in column INTERP_MTHD
ERROR: header information mismatch in column INTERP_MTHD
ERROR: header information mismatch in column INTERP_MTHD
ERROR: failed tests for CNT: hdr
ERROR: failed tests for CTC: hdr
ERROR: failed tests for CTS: hdr
ERROR: failed tests for ECLV#66: hdr
ERROR: failed tests for FHO: hdr
ERROR: failed tests for SL1L2: hdr
ERROR: failed tests for VCNT: hdr
ERROR: failed tests for VL1L2: hdr
file1: MET-feature_1926_gridstat_openmp/test_output/grid_stat/grid_stat_GRIB2_NAM_RTMA_120000L_20120409_120000V_vcnt.txt
file2: MET-develop/test_output/grid_stat/grid_stat_GRIB2_NAM_RTMA_120000L_20120409_120000V_vcnt.txt
ERROR: header information mismatch in column INTERP_MTHD
ERROR: failed tests for VCNT: hdr
file1: MET-feature_1926_gridstat_openmp/test_output/grid_stat/grid_stat_GRIB2_NAM_RTMA_120000L_20120409_120000V_vl1l2.txt
file2: MET-develop/test_output/grid_stat/grid_stat_GRIB2_NAM_RTMA_120000L_20120409_120000V_vl1l2.txt
ERROR: header information mismatch in column INTERP_MTHD
ERROR: failed tests for VL1L2: hdr
```

3. These files are added/modified/fixed by MET#1810
```
ERROR: folder MET-feature_1926_gridstat_openmp/test_output missing 2 files
    tc_gen/tc_gen_2021_shape_pct.txt 
    tc_gen/tc_gen_2021_shape.stat
---
file1: MET-feature_1926_gridstat_openmp/test_output/gen_vx_mask/SHAPE_mask.nc
file2: MET-develop/test_output/gen_vx_mask/SHAPE_mask.nc
ERROR: found    21 differences in var shape_mask                             - max abs: 1 
---
file1: MET-feature_1926_gridstat_openmp/test_output/tc_gen/tc_gen_prob_pjc.txt
file2: MET-develop/test_output/tc_gen/tc_gen_prob_pjc.txt
ERROR: line type PCT#39 not found in stat2
ERROR: line type PJC#55 not found in stat1
ERROR: failed tests for PCT#39: lty
ERROR: failed tests for PJC#55: lty
file1: MET-feature_1926_gridstat_openmp/test_output/tc_gen/tc_gen_prob_prc.txt
file2: MET-develop/test_output/tc_gen/tc_gen_prob_prc.txt
ERROR: line type PCT#39 not found in stat2
ERROR: line type PRC#39 not found in stat1
ERROR: failed tests for PCT#39: lty
ERROR: failed tests for PRC#39: lty
--
file1: MET-feature_1926_gridstat_openmp/test_output/tc_gen/tc_gen_prob.stat
file2: MET-develop/test_output/tc_gen/tc_gen_prob.stat
ERROR: line type PJC#55 not found in stat1
ERROR: line type PRC#39 not found in stat1
ERROR: differing number of rows 9 vs. 3 for row type PCT#39 between versions 10_1 vs. 10_1 
ERROR: failed tests for PJC#55: lty
ERROR: failed tests for PRC#39: lty
ERROR: failed tests for PCT#39: nrow
```
